core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java

```diff
@@ -0,0 +1,357 @@
+
+    public QueryCreationContext(Index index, IndexQueryParserService indexQueryParser) {
+        this.index = index;
+        this.indexVersionCreated = Version.indexCreated(indexQueryParser.indexSettings());
+        this.indexQueryParser = indexQueryParser;
+        this.parseContext = new QueryParseContext(this);
+    }
+
+    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
+        this.parseFieldMatcher = parseFieldMatcher;
+    }
+
+    public ParseFieldMatcher parseFieldMatcher() {
+        return parseFieldMatcher;
+    }
+
+    //norelease the parser argument won't be necessary after complete separation from QueryParseContext
+    public void reset(XContentParser jp) {
+        allowUnmappedFields = indexQueryParser.defaultAllowUnmappedFields();
+        this.parseFieldMatcher = ParseFieldMatcher.EMPTY;
+        this.lookup = null;
+        this.namedQueries.clear();
+        this.nestedScope = new NestedScope();
+        this.parseContext.reset(jp);
+    }
+
+    public Index index() {
+        return this.index;
+    }
+
+    public IndexQueryParserService indexQueryParserService() {
+        return indexQueryParser;
+    }
+
+    public AnalysisService analysisService() {
+        return indexQueryParser.analysisService;
+    }
+
+    public ScriptService scriptService() {
+        return indexQueryParser.scriptService;
+    }
+
+    public MapperService mapperService() {
+        return indexQueryParser.mapperService;
+    }
+
+    @Nullable
+    public SimilarityService similarityService() {
+        return indexQueryParser.similarityService;
+    }
+
+    public Similarity searchSimilarity() {
+        return indexQueryParser.similarityService != null ? indexQueryParser.similarityService.similarity() : null;
+    }
+
+    public String defaultField() {
+        return indexQueryParser.defaultField();
+    }
+
+    public boolean queryStringLenient() {
+        return indexQueryParser.queryStringLenient();
+    }
+
+    public MapperQueryParser queryParser(QueryParserSettings settings) {
+        queryParser.reset(settings);
+        return queryParser;
+    }
+
+    public BitDocIdSetFilter bitsetFilter(Filter filter) {
+        return indexQueryParser.bitsetFilterCache.getBitDocIdSetFilter(filter);
+    }
+
+    public <IFD extends IndexFieldData<?>> IFD getForField(MappedFieldType mapper) {
+        return indexQueryParser.fieldDataService.getForField(mapper);
+    }
+
+    public void addNamedQuery(String name, Query query) {
+        namedQueries.put(name, query);
+    }
+
+    public ImmutableMap<String, Query> copyNamedQueries() {
+        return ImmutableMap.copyOf(namedQueries);
+    }
+
+    public void combineNamedQueries(QueryCreationContext context) {
+        namedQueries.putAll(context.namedQueries);
+    }
+
+//    /**
+//     * Return whether we are currently parsing a filter or a query.
+//     */
+//    public boolean isFilter() {
+//        return isFilter;
+//    }
```

Comment:
I'm unsure, was hoping we only need this in fromXContent(), thats why I removed it from this context. However, we might need something similar so this when traversing the query builder tree when generating the lucene query. Something on my todo list to investigate.